eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState/DPState.hs

```diff
@@ -235,19 +268,166 @@ instance Default (DState c) where
       Map.empty
       (GenDelegs Map.empty)
       def
+      Map.empty
 
 instance Default (PState c) where
   def =
-    PState Map.empty Map.empty Map.empty
+    PState Map.empty Map.empty Map.empty Map.empty
 
 rewards :: DState c -> ViewMap c (Credential 'Staking c) Coin
-rewards (DState unified _ _ _) = Rewards unified
+rewards (DState unified _ _ _ _) = Rewards unified
 
 delegations ::
   DState c ->
   ViewMap c (Credential 'Staking c) (KeyHash 'StakePool c)
-delegations (DState unified _ _ _) = Delegations unified
+delegations (DState unified _ _ _ _) = Delegations unified
 
 -- | get the actual ptrs map, we don't need a view
 ptrsMap :: DState c -> Map Ptr (Credential 'Staking c)
-ptrsMap (DState (UnifiedMap _ ptrmap) _ _ _) = ptrmap
+ptrsMap (DState (UnifiedMap _ ptrmap) _ _ _ _) = ptrmap
+
+-- ==========================================================
+-- Functions that handle Deposits for stake credetials and key hashes.
+
+-- | One only pays a deposit on the initial key registration. If the key has been
+--   de-registered it should have been removed from the map. If it hasn't been
+--   de-registered, then it has no effect on the Deposits. In places where this function
+--   is called, there should be an explicit check that the credential is not in the map.
+payKeyDeposit ::
+  HasField "_keyDeposit" (PParams era) Coin =>
+  Credential 'Staking (EraCrypto era) ->
+  PParams era ->
+  DState (EraCrypto era) ->
+  DState (EraCrypto era)
+payKeyDeposit cred pp dstate = dstate {dsDeposits = newStake}
+  where
+    stake = dsDeposits dstate
+    newStake = case Map.lookup cred stake of
+      Nothing -> Map.insert cred (getField @"_keyDeposit" pp) stake
+      Just _ -> stake
```

Comment:
I am pretty sure yor suggestion has it backwards

 ```
newStake = case Map.lookup cred stake of
      Nothing -> Map.insert cred (getField @"_keyDeposit" pp) stake
      Just _ -> stake
```

Should be recast as

 ```
newState | Map.notMember cred stake = Map.insert cred (getField @"_keyDeposit" pp) stake
             | otherwise = state
```

Reply:
Yeah, you are right! That's exactly what I meant to say :wink: